### PR TITLE
4196 refactor chips

### PR DESCRIPTION
**Pull request by Liliya-web**

### CODE REVIEW

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 1]
```diff
@@ -15,7 +19,8 @@
  * To see an example of Chip web element please visit https://vuetifyjs.com/en/components/chips/
  */
 
-public class Chip extends UIBaseElement<ChipAssert> implements HasClick, HasIcon, HasLabel {
+public class Chip extends UIBaseElement<ChipAssert> implements HasClick, HasLabel, HasColor, HasIcon, HasImage,
+        HasTheme, IsOutlined {
 
     private static final String TEXT1 = ".v-chip__content";
```
Comment (omega790) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:25]: It's just custom implementation, please delete it from framework:
ALL_TEXTS_FROM_COMPOSITE_LABEL
BOLD_TEXT_FROM_COMPOSITE_LABEL
REGULAR_TEXT_FROM_COMPOSITE_LABEL

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 2]
```diff
@@ -40,17 +51,17 @@ public String getText() {
         }
     }
 
-    @JDIAction("Get bold text from {name}'s composite label")
+    @JDIAction("Get bold text from '{name}' composite label")
     public String getBoldTextFromCompositeLabel() {
```
Comment (omega790) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:55]: custom implementation -> delete

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 3]
```diff
@@ -40,17 +51,17 @@ public String getText() {
...
         return find(BOLD_TEXT_FROM_COMPOSITE_LABEL).getText();
     }
 
-    @JDIAction("Get bold text from {name}'s composite label")
+    @JDIAction("Get bold text from '{name}' composite label")
     public String getRegularTextFromCompositeLabel() {
```
Comment (omega790) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:60]: custom implementation -> delete

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 4]
```diff
@@ -40,17 +51,17 @@ public String getText() {
...
         return find(REGULAR_TEXT_FROM_COMPOSITE_LABEL).getText();
     }
 
-    @JDIAction("Get all the text from {name}'s composite label")
+    @JDIAction("Get all the text from '{name}' composite label")
     public String getFullTextFromCompositeLabel() {
```
Comment (omega790) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:65]: custom implementation -> delete

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 5]
```diff
@@ -93,32 +104,49 @@ public boolean hasImage() {
             .wait(() -> this.find(IMAGE).isExist());
     }
 
-    @JDIAction("Get {name}'s height")
-    public String height() {
-        return css("height");
+    @JDIAction("Get '{name}' height")
```
Comment (omega790) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:107]: Please use HasMeasurement interface

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 6]
```diff
@@ -93,32 +104,49 @@ public boolean hasImage() {
...
+    public boolean hasLargeSize() {
+        return hasClass("v-size--large");
     }
 
-    @JDIAction("Get {name}'s borderColor")
+    @JDIAction("Check that '{name}' has x-large size")
+    public boolean hasXLargeSize() {
+        return hasClass("v-size--x-large");
+    }
+
```
Comment (omega790) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:141]: excessive two empty lines

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/asserts/ChipAssert.java [hunk 7]
```diff
@@ -3,131 +3,195 @@
...
-
-    //For simple (not composite) label
-    @JDIAction("Assert that {name}'s label contains '{0}'")
+public class ChipAssert extends UIAssert<ChipAssert, Chip> implements ColorAssert<ChipAssert, Chip>,
+        OutlinedAssert<ChipAssert, Chip>, ThemeAssert<ChipAssert, Chip> {
+    /**
+     * For simple (not composite) label
+     */
+    @JDIAction("Assert that '{name}' label contains text '{0}'")
     public ChipAssert containsText(String text) {
```
Comment (omega790) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/asserts/ChipAssert.java:19]: Please use common methods from ITextAssert

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/asserts/ChipAssert.java [hunk 8]
```diff
@@ -3,131 +3,195 @@
...
     @JDIAction("Assert that '{name}' has image")
     public ChipAssert image() {
-        jdiAssert(element().hasImage() ? "has image" : "has no image", Matchers.is("has image"));
+        jdiAssert(element().hasImage(), Matchers.is(true), "Element has not image");
         return this;
     }
 
-    @JDIAction("Assert that '{name}' has height {0} px")
+    @JDIAction("Assert that '{name}' has height '{0} px'")
     public ChipAssert height(int height) {
```
Comment (omega790) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/asserts/ChipAssert.java:129]: Please use MeasurementAssert interface

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java [hunk 9]
```diff
@@ -115,22 +143,49 @@ public void customListChipTests() {
...
-
         expandableChip.click();
         waitCondition(() -> expandableMenu.isDisplayed());
         assertThat(expandableMenu.isDisplayed(), Matchers.is(true));
         expandableMenuItems.get(1).click(); // collapse menu just to be safe
     }
 
     @Test
-    public void inSelectChipTests() {
+    public void labelTextChipTests() {
```
Comment (omega790) [inline jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java:153]: This test for custom realization -> I suppose we don't need it

File: jdi-light-vuetify-tests/src/main/java/io/github/com/pages/ChipsPage.java [hunk 10]
```diff
@@ -49,24 +41,15 @@ public class ChipsPage extends VuetifyPage {
...
-    public static Button customListNextButton;
-
     @UI("#ExpandableChip .v-chip")
     public static Chip expandableChip;
 
-    @UI(".v-card .v-list-item__subtitle")
-    public static List<Text> expandableChipEmails;
-
     @UI("div[role='menu']")
     public static UIElement expandableMenu;
```
Comment (omega790) [inline jdi-light-vuetify-tests/src/main/java/io/github/com/pages/ChipsPage.java:54]: I suppose this element is Menu, so in test we can use Menu's methods

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java [hunk 11]
```diff
@@ -115,22 +143,49 @@ public void customListChipTests() {
     public void expandableChipTests() {
         assertThat(expandableMenu.isNotDisplayed(), Matchers.is(true));
```
Comment (omega790) [inline jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java:144]: Use Menu's methods

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java [hunk 12]
```diff
@@ -115,22 +143,49 @@ public void customListChipTests() {
...
         expandableChip.has().image();
-
         expandableChip.click();
         waitCondition(() -> expandableMenu.isDisplayed());
         assertThat(expandableMenu.isDisplayed(), Matchers.is(true));
```
Comment (omega790) [inline jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java:148]: Use Menu's methods

**Review by omega790 — CHANGES_REQUESTED** [review]

File: jdi-light-vuetify-tests/src/main/java/io/github/com/pages/ChipsPage.java (hunk 10 above)
Comment (pnatashap) [reply jdi-light-vuetify-tests/src/main/java/io/github/com/pages/ChipsPage.java:54]: Menu is available for all pages, so it should be a part of VuetifyPage

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java [hunk 13]
```diff
@@ -41,67 +47,89 @@ public void before() {
...
+        closableChip.is().displayed().and().removable();
         closableChip.close();
         closableChip.is().hidden();
+        greyChip.show();
+        greyChip.is().notRemovable();
     }
 
     @Test
     public void coloredChipTests() {
-        coloredChips.get(1).has().backgroundColor(Colors.GREY_LIGHTEN_2.value());
```
Comment (pnatashap) [inline jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java:52]: The test name is coloredChipTests, but we are working with outlinedChip.  It is not very good. 

The aim of this project is also to be documentation and example, how to work with elements. 
So it should be easy to find correct examples using the test names

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 14]
```diff
@@ -31,7 +36,13 @@ public ChipAssert is() {
...
+    @Override
+    @JDIAction("Check that '{name}' is enabled")
+    public boolean isEnabled() {
+        return !hasClass("v-chip--disabled");
+    }
+
+    @JDIAction("Get '{name}' text")
     public String getText() {
         if (!find(TEXT1).getText().isEmpty()) {
             return find(TEXT1).getText();
```
Comment (pnatashap) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:48]: Are we sure that we need this if/else? As I see in both cases we have .v-chip__content and should get the text.

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 15]
```diff
@@ -15,7 +19,8 @@
  * To see an example of Chip web element please visit https://vuetifyjs.com/en/components/chips/
  */
 
-public class Chip extends UIBaseElement<ChipAssert> implements HasClick, HasIcon, HasLabel {
+public class Chip extends UIBaseElement<ChipAssert> implements HasClick, HasLabel, HasColor, HasIcon, HasImage,
```
Comment (pnatashap) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:22]: As I understand we can create a custom content for the chip. In this case we should be able to create a custom class to describe the content and get ability to get this content as UIElement or cast it to specific class. 

Ask me directly if you have any questions or need an example

**Review by pnatashap — CHANGES_REQUESTED** [review]

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java (hunk 15 above)
Comment (vklonin) [reply jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:22]: Let's have a close look in to this!

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java (hunk 9 above)
Comment (Liliya-web) [reply jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java:153]: As Natalia suggested, I moved custom realization to custom folder in tests module

File: jdi-light-vuetify-tests/src/main/java/io/github/com/pages/ChipsPage.java (hunk 10 above)
Comment (Liliya-web) [reply jdi-light-vuetify-tests/src/main/java/io/github/com/pages/ChipsPage.java:54]: It's a menu, which opens when we click on the expandable chip

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java (hunk 14 above)
Comment (Liliya-web) [reply jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:43]: It seems that it works with TEXT1 even without condition, thanks!

File: jdi-light-vuetify-tests/src/main/java/io/github/com/custom/CompositeLabelChip.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
+package io.github.com.custom;
+
+import com.epam.jdi.light.common.JDIAction;
+import com.epam.jdi.light.elements.base.UIBaseElement;
+import com.epam.jdi.light.elements.common.UIElement;
+import io.github.com.custom.asserts.CompositeLabelChipAssert;
+
+import java.util.List;
+
+public class CompositeLabelChip extends UIBaseElement<CompositeLabelChipAssert> {
```
Comment (omega790) [inline jdi-light-vuetify-tests/src/main/java/io/github/com/custom/CompositeLabelChip.java:10]: CompositeLabelChip should be extension for Chip (we use common methods + add methods for customization)
For example, now we can't use methods like selected()/deselected() and others

**Review by omega790 — CHANGES_REQUESTED** [review]

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 17]
```diff
@@ -93,54 +83,53 @@ public boolean hasImage() {
...
-        return !css("border-color").equals(css("background-color"))
-            && !css("border-color").contains("transparent");
+    @JDIAction("Get '{name}' borderColor")
+    public String borderColor() {
+        return css("border-color");
     }
 
     @Override
     public void click() {
         core().click(1, getSize().getHeight() / 2); //click left center in order to miss a closing button if it is on a chip
```
Comment (pnatashap) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:123]: I t is better to click on text or content if there are any, if not then center

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 18]
```diff
@@ -31,34 +35,20 @@ public ChipAssert is() {
...
+    @JDIAction("Check that '{name}' is enabled")
+    public boolean isEnabled() {
+        return !hasClass("v-chip--disabled");
     }
 
-    @JDIAction("Get bold text from {name}'s composite label")
-    public String getRegularTextFromCompositeLabel() {
-        return find(REGULAR_TEXT_FROM_COMPOSITE_LABEL).getText();
+    @JDIAction("Get '{name}' content")
+    public List<UIElement> getContent() {
```
Comment (pnatashap) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:45]: chip content is always one, as I can see (do you have any examples for several?)
so our locator should for element ".v-chip__content" and after that it can be customized. 
It is not possible to automatically cast the list of some elements to several another classes. That's why we need one element which contains some other elements. For example for "Expandable chip" on owr site we can create a custom class with image and name

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 19]
```diff
@@ -76,13 +66,13 @@ public boolean selected() {
         return hasClass("v-chip--active");
```
Comment (pnatashap) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:66]: Comment and check are equal, but method name is about selection, should be renamed

**Review by pnatashap — CHANGES_REQUESTED** [review]

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java (hunk 18 above)
Comment (Liliya-web) [reply jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:45]: Chip content may contain text, icons, images, and avatars. Do you mean we should create some more custom classes for all that cases?

Comment (Liliya-web) [reply jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:45]: We can get and cast elements using stream()
I put an example in tests

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java [hunk 20]
```diff
@@ -31,34 +32,25 @@ public ChipAssert is() {
...
+    @JDIAction("Get '{name}' content")
+    public UIElement getContent() {
+        return find(CONTENT);
     }
 
-    @JDIAction("Get bold text from {name}'s composite label")
-    public String getRegularTextFromCompositeLabel() {
-        return find(REGULAR_TEXT_FROM_COMPOSITE_LABEL).getText();
+    @JDIAction("Get '{name}' text")
+    public Text text() {
```
Comment (pnatashap) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/common/Chip.java:47]: Text type is not useful here

**Review by pnatashap — CHANGES_REQUESTED** [review]

File: jdi-light-vuetify-tests/src/main/java/io/github/com/custom/chips/CompositeLabelChip.java [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+	public List<Text> compositeLabel() {
+		return finds(ALL_TEXTS_FROM_COMPOSITE_LABEL).stream()
+				.map(e -> new Text().setCore(Text.class, e))
+				.collect(Collectors.toList());
+	}
+
+	@JDIAction("Get all the text from '{name}' composite label")
+	public String getFullTextFromCompositeLabel() {
+		StringBuilder result = new StringBuilder();
+		for (int i = 0; i <= compositeLabel().size() - 1; i++) {
```
Comment (pnatashap) [inline jdi-light-vuetify-tests/src/main/java/io/github/com/custom/chips/CompositeLabelChip.java:41]: List<Text> labels = compositeLabel();
And after that use labels for iterations.

File: jdi-light-vuetify-tests/src/main/java/io/github/com/custom/chips/CompositeLabelChip.java [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+			result.append(compositeLabel().get(i).getText());
+			result.append(" ");
```
Comment (pnatashap) [inline jdi-light-vuetify-tests/src/main/java/io/github/com/custom/chips/CompositeLabelChip.java:43]: Not a good solution, as we will get additional space at the end of the line. So it is better to create an array of string and after that use String.join

File: jdi-light-vuetify-tests/src/main/java/io/github/com/enums/Colors.java [hunk 23]
```diff
@@ -85,6 +85,7 @@ public enum Colors {
     BLUE_ACCENT_2("rgba(68, 138, 255, 1)"),
     BLUE_ACCENT_3("rgba(41, 121, 255, 1)"),
```
Comment (pnatashap) [inline jdi-light-vuetify-tests/src/main/java/io/github/com/enums/Colors.java:86]: Looks like that color names are used by vuetify, so may be we can add them to vuetify package. Can be a separate PR, please check this.

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java [hunk 24]
```diff
@@ -38,99 +45,158 @@ public void before() {
...
+        outlinedChip.is().outlined();
+        sizeChip.is().notOutlined();
     }
 
-    @Test
+    @Test (description = "Text checks chip sizes")
     public void sizesChipTests() {
         int expectedFontSize = 10;
         int expectedHeight = 16;
+        sizesChips.get(1).show();
```
Comment (pnatashap) [inline jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java:111]: Chip firstChip = sizesChips.get(1); and after that use it to check all

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java [hunk 25]
```diff
@@ -38,99 +45,158 @@ public void before() {
...
         sizesChips.get(1).has().fontSize(expectedFontSize);
         sizesChips.get(1).has().height(expectedHeight);
+        sizesChips.get(1).has().xSmallSize();
+        sizesChips.get(2).has().smallSize();
```
Comment (pnatashap) [inline jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/common/ChipsTests.java:115]: If the indexed is used only once, we can use without variable

**Review by pnatashap — CHANGES_REQUESTED** [review]

**Review by pnatashap — APPROVED** [review]
